Subject: nfsd: support disabling 64bit dir cookies
References: bnc#937503
From: NeilBrown <neilb@suse.com>
Patch-mainline: never, compatability support for customers

A recent sles bugfix causes nfsd to use 64bit directory
cookies with ext3.
This is needed for correct handling of large directories, but
some NFSv3 clients are buggy and don't handle 64bit cookies
correctly.
So this change is a regression for customers with those buggy
clients.

So add a module parameter to restore the old behavior.
Setting
   nfsd.allow_large_cookies=N
will resort the old bechaviour for all exported filesystems.
This can be acheived by
  echo N > /sys/modules/nfsd/parameters/allow_large_cookies

Signed-off-by: NeilBrown <neilb@suse.com>

---
 fs/nfsd/vfs.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/fs/nfsd/vfs.c
+++ b/fs/nfsd/vfs.c
@@ -2012,6 +2012,11 @@ static __be32 nfsd_buffered_readdir(stru
 	return cdp->err;
 }
 
+#include <linux/module.h>
+static bool allow_large_cookies = true;
+module_param(allow_large_cookies, bool, 0644);
+MODULE_PARM_DESC(allow_large_cookies,
+		 "Allow NFS server to send 64 bit readdir cookies to clients.");
 /*
  * Read entries from a directory.
  * The  NFSv3/4 verifier we ignore for now.
@@ -2026,7 +2031,7 @@ nfsd_readdir(struct svc_rqst *rqstp, str
 	int             may_flags = NFSD_MAY_READ;
 
 	/* NFSv2 only supports 32 bit cookies */
-	if (rqstp->rq_vers > 2)
+	if (rqstp->rq_vers > 2 && allow_large_cookies)
 		may_flags |= NFSD_MAY_64BIT_COOKIE;
 
 	err = nfsd_open(rqstp, fhp, S_IFDIR, may_flags, &file);
